webui/react/src/stores/experiments.tsx

```diff
@@ -1,63 +1,96 @@
 import { Map } from 'immutable';
-import React, { createContext, PropsWithChildren, useCallback, useContext, useState } from 'react';
 
 import { getExperiments } from 'services/api';
 import { V1Pagination } from 'services/api-ts-sdk';
 import { GetExperimentsParams } from 'services/types';
-import { ExperimentItem } from 'types';
+import { ExperimentItem, ExperimentPagination } from 'types';
 import handleError from 'utils/error';
 import { Loadable, Loaded, NotLoaded } from 'utils/loadable';
+import { Observable, observable, WritableObservable } from 'utils/observable';
 import { encodeParams } from 'utils/store';
 
-type ExperimentPagination = {
-  experiments: ExperimentItem[];
-  pagination?: V1Pagination;
+type ExperimentsCache = {
+  experimentIds: Readonly<number[]>;
+  pagination: Readonly<V1Pagination>;
 };
 
-type ExperimentsContext = {
-  experimentsCache: Map<string, ExperimentPagination>;
-  updateExperimentsCache: (ws: Map<string, ExperimentPagination>) => void;
-};
+// TODO: use js private intead of ts private for methods
+export class ExperimentsService {
+  #experimentsCache: WritableObservable<Map<string, ExperimentsCache>> = observable(Map());
+  #experimentMap: WritableObservable<Map<number, ExperimentItem>> = observable(Map());
 
-const ExperimentsContext = createContext<ExperimentsContext | null>(null);
-
-export const ExperimentsProvider: React.FC<PropsWithChildren> = ({ children }) => {
-  const [experimentsCache, updateExperimentsCache] = useState<Map<string, ExperimentPagination>>(
-    Map<string, ExperimentPagination>(),
-  );
-  return (
-    <ExperimentsContext.Provider value={{ experimentsCache, updateExperimentsCache }}>
-      {children}
-    </ExperimentsContext.Provider>
-  );
-};
+  // Get an experiment by experiment id
+  public getExperimentsByIds(experimentIds: number[]): Observable<Loadable<ExperimentItem[]>> {
+    return this.#experimentMap.select((map) => {
+      const expList: ExperimentItem[] = experimentIds.flatMap((id) => {
+        const exp = map.get(id);
+        return exp ? [exp] : [];
+      });
+      return Loaded(expList);
+    });
+  }
 
-export const useFetchExperiments = (
-  params: GetExperimentsParams,
-  canceler: AbortController,
-): (() => Promise<void>) => {
-  const context = useContext(ExperimentsContext);
-  if (context === null) {
-    throw new Error('Attempted to use useFetchExperiments outside of Experiment Context');
+  public getExperimentsByParams(
```

Comment:
I think `experiments` in the method names is a little redundant, you could shorten these if you want.

Reply:
i'll leave as it is for now.